core/src/test/java/com/linecorp/armeria/server/ServerBuilderTest.java

```diff
@@ -684,4 +684,157 @@ void exceptionReportIntervalMilliSeconds() {
                       .service("/", (ctx, req) -> HttpResponse.of(HttpStatus.OK))
                       .build());
     }
+
+    @Test
+    void virtualHostContextPath() {
+        final Server server = Server
+                .builder()
```

Comment:
I think we should also consider decorators:

```java
Server.builder()
  .decorator("/asdf", (delegate, ctx, req) -> HttpResponse.of(200))
...
```

We probably need to apply the `contextPath` like the following:

```
@@ -737,7 +737,7 @@ public final class VirtualHostBuilder implements TlsSetters {
 
     @Nullable
     private Function<? super HttpService, ? extends HttpService> getRouteDecoratingService(
-            @Nullable VirtualHostBuilder defaultVirtualHostBuilder) {
+            @Nullable VirtualHostBuilder defaultVirtualHostBuilder, String contextPath) {
         final List<RouteDecoratingService> routeDecoratingServices;
         if (defaultVirtualHostBuilder != null) {
             routeDecoratingServices = ImmutableList.<RouteDecoratingService>builder()
@@ -749,8 +749,11 @@ public final class VirtualHostBuilder implements TlsSetters {
         }
 
         if (!routeDecoratingServices.isEmpty()) {
-            return RouteDecoratingService.newDecorator(
-                    Routers.ofRouteDecoratingService(routeDecoratingServices));
+            final List<RouteDecoratingService> prefixed =
+                    routeDecoratingServices.stream()
+                                           .map(service -> service.withRoutePrefix(contextPath))
+                                           .collect(Collectors.toList());
+            return RouteDecoratingService.newDecorator(Routers.ofRouteDecoratingService(prefixed));
         } else {
             return null;
         }
```

where the routes can be overridden using delegation
```
+++ b/core/src/main/java/com/linecorp/armeria/internal/server/RouteDecoratingService.java
@@ -91,6 +91,15 @@ public final class RouteDecoratingService implements HttpService {
         decorator = requireNonNull(decoratorFunction, "decoratorFunction").apply(this);
     }
 
+    public RouteDecoratingService(Route route, HttpService decorator) {
+        this.route = requireNonNull(route, "route");
+        this.decorator = requireNonNull(decorator, "decorator");
+    }
+
+    public RouteDecoratingService withRoutePrefix(String prefix) {
+        return new RouteDecoratingService(route.withPrefix(prefix), decorator);
+    }
+
```

Reply:
Ah... decorator. thanks.